text/0000-dll-kind.md

```diff
@@ -0,0 +1,103 @@
+#[allow(non_snake_case)]
+extern "system" {
+    fn GetStdHandle(nStdHandle: u32) -> *mut u8;
+}
+```
+
+Some symbols are only exported by ordinal from the dll in which case `#[link_ordinal]` may be used:
+
+```rust
+#[cfg(windows)]
+#[link(name = "ws2_32.dll", kind = "dll")]
+#[allow(non_snake_case)]
+extern "system" {
+    #[link_ordinal(116)]
+    fn WSACleanup() -> i32;
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Add a new attribute `#[link_ordinal]` taking a single numerical value, such as `#[link_ordinal(116)]`. It can only be specified on symbols in an extern block using `kind="dll"`.
+
+Add a new value `dll` to the `kind` property of the `link` attribute. When this kind is specified, the `name` must explicitly include the extension. In addition, for all items in the associated extern block Rust will *keep* the symbol mangled, instead of having an unmangled symbol. Rust will emit an idata section that maps from the *mangled* symbol to a symbol in the specified dll. The symbol in the dll that the idata section maps to depends on which attributes are specified on the item in question:
+
+* If `#[link_ordinal]` is specified the idata section will map from the mangled symbol to the ordinal specified in the dll.
+* If `#[link_name]` is specified the idata section will map from the mangled symbol to the name specified in the dll, without any calling convention decorations added. If calling convention decorations are desired they must be specified explicitly in the value of the `#[link_name]` attribute.
+* If both `#[link_ordinal]` and `#[link_name]` are specified, an error will be emitted.
+* If neither `#[link_ordinal]` nor `#[link_name]` are specified, the idata section will map from the mangled symbol to its unmangled equivalent in the dll. The unmangled symbol will *not* have calling convention decorations.
+
+The idata section that is produced is equivalent to the idata sections found in import libraries, and should result in identical code generation by the linker.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Additional complexity in the language through a new `kind` and a new attribute for specifying ordinals.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+The RFC as proposed would allow for full control over linking to symbols from dlls with syntax as close as possible to existing extern blocks.
+
+No alternatives are currently known other than the status quo.
+
+# Prior art
+[prior-art]: #prior-art
+
+Many non-native languages have the ability to import symbols from dlls, but this uses runtime loading by the language runtime and is not the same as what is being proposed here.
+
+Delphi is a native language that has the ability to import symbols from dlls without import libraries.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+* Bikeshedding on attribute names.
+* Should this feature be extended to other platforms?
```

Comment:
Being able to link to shared libraries on other platforms without actually having the shared library around at link time. I don't really *know* if that is feasible or desired.

Reply:
You might want to move the question over to "future possibilities" with such a note / speculation since it isn't unresolved for the purposes of stabilization & the tracking issue.

Reply:
It would be a question of whether the name `dll` for the `kind` would be appropriate if it works for shared libraries other than dlls.

Reply:
So what would we call if it was for other shared libs? `kind = "shared"` perhaps?

Reply:
Please do drop these two entries in favor of the paragraph I suggested below in the future possibilities section, as well as changing the attribute name to `kind = "shared"` or `kind = "dylib"` (your call, let's not bikeshed this).